File: ge1400/LT1493_Longest_Subarray_of_1s_After_Deleting_One_Element.cpp

#include "../header/myheader.h"

class LT1493
{
public:

// D D


//        int i = 0, j, k = 1;
//        for (j = 0; j < A.size(); ++j) {
//            if (A[j] == 0) k--;
//            if (k < 0 && A[i++] == 0) k++;
//        }
//        return j - i - 1;
// 窗口滑动， 还有之前的  只大不小， 就是不用while，而是if。


//    int max_sz = 0;
//    for (int i = 0, j = 0, skip = 0; i < nums.size(); ++i) {
//        skip += nums[i] == 0;
//        if (skip > 1)
//            skip -= nums[j++] == 0;
//        max_sz = max(max_sz, i - j + 1);
//    }
//    return max_sz - 1;


//Runtime: 104 ms, faster than 11.38% of C++ online submissions for Longest Subarray of 1's After Deleting One Element.
//Memory Usage: 36.8 MB, less than 84.04% of C++ online submissions for Longest Subarray of 1's After Deleting One Element.
    int lt1493a(vector<int>& nums)
    {
        int llst0 = -1;
        int lst0 = -1;
        int ans = 0;
        for (int i = 0; i < nums.size(); ++i)
        {
            if (nums[i] == 0)
            {
                ans = max(ans, i - llst0 - 2);
//                cout<<ans<<", "<<i<<", "<<llst0<<" , "<<lst0<<endl;
                llst0 = lst0;
                lst0 = i;
            }
        }
//        cout<<ans<<endl;
        ans = max(ans, int (nums.size()) - llst0 - 2);
        return ans;
    }

};

int main()
{

//    myvi v = {1,1,0,1};
//    myvi v = {0,1,1,1,0,1,1,0,1};
//    myvi v = {1,1,1};
//    myvi v = {0,0,0};
    myvi v = {1,1,0,0,1,1,1,0,1};

    LT1493 lt;

    cout<<lt.lt1493a(v)<<endl;

    return 0;
}

File: other_language/LT1493_longest-subarray-of-1s-after-deleting-one-element.rb







# ren hao shao
# Runtime: 100 ms, faster than 100.00% of Ruby online submissions for Longest Subarray of 1's After Deleting One Element.
# Memory Usage: 213 MB, less than 100.00% of Ruby online submissions for Longest Subarray of 1's After Deleting One Element.
# @param {Integer[]} nums
# @return {Integer}
def longest_subarray(nums)
    ans = 0;
    len0 = 0;
    len1 = 0;
    lst1 = 0;
    lst0 = 0;
    nums.each do |val|

        if val == 1 then
            len1 += 1
            if len1 == 1 then 
                lst0 = len0;
                len0 = 0;
            end
            if len1 > ans then
                ans = len1
            end
        else
            # ans = max()       // where is max ?
            if lst0 == 1 then
                if lst1 + len1 > ans then
                    ans = lst1 + len1;
                end
            end
            len0 += 1
            if (len0 == 1) then
                lst1 = len1;
                len1 = 0;
            end                        # 可以增加：len0 > 1, lst1 = 0;
        end                 # 吐了， if then end，匹配不起来。。一直说这里多了一个end， 然后 44行(就是最后一行) 少了一个end
                        # ruby 好像没有++，， len0++，不知道翻译成什么了。。。。导致匹配不上
    end

    if (lst0 == 1 and (lst1 + len1 > ans)) then
        ans = lst1 + len1;
    end
    if (len1 > ans) then ans = len1 end
    if (lst1 > ans) then ans = lst1 end
    # if ans == nums.size then ans -= 1 end
    if lst0 == 0 && len0 == 0 then ans = nums.length - 1 end

    return ans;
end

puts "hi"

# arr = Array[1,1,0,1]
# arr = Array[0,1,1,1,0,1,1,0,1]
# arr = Array[1,1,1,0]
# arr = Array[1,1,0,0,1,1,1,0,1]
# arr = Array[0,0,0]
# 14
arr = Array[1,0,0,1,1,1,1,1,0,0,0,0,0,0,1,1,1,0,1,1,1,1,1,0,1,1,1,1,1,1,1,1,0,1,1,0,0,0,1,1,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,1,1,1,1,0,1,1]

ans = longest_subarray(arr);

puts ans
